### PR TITLE
fix(state-inspector): a link's schema renders as what the link stores

### PR DESCRIPTION
**Pull request by mathpirate**
## The defect

`cf inspect value-at` reported a stored link's schema as `true`:

```json
{"$link": {"id": "of:fid1:1D7eZcK…", "scope": "space", "schema": true}}
```

What the store actually holds on that link is a ~2.2KB object: `$defs`,
`properties`, a `required` list of 13, `asCell` stream declarations, and an
external `$ref`. Across all 141 revisions of that document the literal
`"schema":true` appears zero times — the `true` was manufactured on the read
path.

This is worse than lossy. A stored schema is a JSON Schema, so `true` is a
value a link can genuinely hold, and it is the one that selects every value —
a link storing it declares no constraint at all. So a schema that pins a
13-field shape rendered as the value meaning the opposite, indistinguishably
from a link that really stores `true`.

## Where it was

One place, at decode time rather than at display time. `DecodedLink` kept only
`hasSchema: boolean`, so the schema was gone before anything could render it,
and `annotate()` then wrote `schema: true` from that boolean.

Every command that renders a value through `annotate()` shared it: `value-at`,
`piece`, `html`, `overlay`, `diff`, `timeline`, `converge`. `entities` and
`graph` go through `summarizeLink()` instead, which prints `+schema` — terse,
but it never fabricated a value.

`diff` and `timeline` had a second symptom from the same cause: both sides of a
comparison rendered `true`, so a revision that *changed* a link's schema showed
no change at all.

## The fix

`DecodedLink` carries the stored schema. `annotate()` renders it:

- **absent** → no `schema` key, as before
- **small** (≤200 bytes of JSON) → the schema itself, so `true`, `false`, and
  `{}` each read back as themselves
- **larger** → no `schema` key, and a sibling `$schemaSummary: { keys, bytes,
  digest }` on `$link`

The summary is deliberate — these schemas are genuinely huge and would drown
the value carrying them — but it sits **beside** `schema` rather than in it.
That placement is the whole point: a link can store a schema of any shape, so
a summary written into the `schema` slot could be a schema some link really
holds, and the rendering would be forgeable in exactly the way the `true` bug
was. `payloadToLink()` bounds what a stored link contributes to `id`, `space`,
`path`, `scope`, and `schema`, so nothing stored can reach a `$`-prefixed
sibling.

The digest makes the summary self-sufficient for the question that actually
gets asked: on the repro document all seven schema-bearing links share digest
`dP-TGNZVNhod`. Different digests prove two schemas differ; equal ones make
agreement overwhelmingly likely without proving it, since the hash is
truncated. It also restores `diff`/`timeline`'s ability to see a schema change
at all.

`--full-depth` writes every schema out in full. It is the annotated form, not
the stored bytes: a schema is walked like any other value, so a sigil-shaped
literal under `const` / `default` / `enum` reads back as `$link` / `$ref`. That
is deliberate — the stored form is not JSON-safe. A `bigint` in a schema breaks
`JSON.stringify()` outright and throws from `stringifyInspectorJson()`; a
`FabricLink` in one flattens to `{}` and disappears, the exact vanishing act
`decodedLinkOf()` exists to prevent. The digest hashes the *stored* schema, so
that is what to compare two links by. (For the repro link, which holds no such
literal, full-depth output is byte-identical to what SQLite holds.)

### `detail.ts` — the same defect, twice more

`buildAllDetails()` indexed a piece's children with `parseSigilLink()`, so a
child named through a modern `FabricLink` was never named, never labeled, and
never reached the declared-schema lookup. `linksWithPaths()` had the same gap
in the outgoing-link list. Both read the link in either at-rest form now.

`declaredSchemaFor()` accepted only an object schema off the naming link, so a
link storing `true` or `false` fell through to the owner's declaration —
reporting a constraint the link does not carry, in place of the one value that
says it carries none. Any stored schema answers now; only an object one has
keys to list.

### `--json` compatibility

The in-repo consumers of `cf inspect --json` are `scripts/topics-export.ts`
(reads `$link.id`) and `scripts/topics-rehearsal-lib.ts` (tests for the
presence of `$link`). Neither reads `$link.schema`; the `scripts` suite passes.

## Verification

Reproduced against the real snapshot before and after, including that entries
7..105 of the same array still render with no `schema` key.

The tests were validated by sabotage: restoring `schema: true` fails 8 of the
10 `annotate()` cases. The two that stay green are the absent case and the
stored-`true` case — precisely the two the old code got right, which is why the
bug was invisible.

The `detail.ts` fixes are pinned the same way: reverting either fails its test
and leaves the two unchanged-behavior cases green.

Gates: `deno fmt --check`, `deno lint`, `deno task check`,
`deno task check-skill-facts`, and the `state-inspector` (105), `cli` (175),
and `scripts` (5) suites all green.

Coverage: the ratchet caught a real `+1` in `packages/cli` — the continuation
line of a concatenated help string, which V8 attributes no range to. Measured
locally to confirm (3194 → 3193 with the string back on one line), so the group
returns to baseline. `packages/state-inspector` goes 465 → 454 on the same
harness, the new `detail.ts` tests covering lines that had none.

Docs updated: `skills/state-inspector/SKILL.md` (honesty contract) and
`packages/state-inspector/README.md` (ground truth vs. hint).

## Not duplicated here

- #6049 — silent truncation in `cf inspect entities`
- `fable/array-read-schema-link` — link resolution for schemas that constrain
  nothing (the runtime behavior this rendering was misreporting)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<a href="https://cubic.dev/pr/commontoolsinc/labs/pull/6063?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/state-inspector/decode.ts [hunk 1]
```diff
@@ -197,10 +204,65 @@ interface AnnotationLeave {
...
+ * otherwise. Never synthesizes a schema — a rendered `true` means `true` was
+ * what the link stored.
+ */
+function renderLinkSchema(schema: Json, maxDepth: number): Json {
+  const rendered = annotate(schema, Number.POSITIVE_INFINITY);
+  if (!Number.isFinite(maxDepth)) return rendered;
+  const bytes = utf8.encode(JSON.stringify(rendered)).length;
+  return bytes <= MAX_INLINE_SCHEMA_BYTES
+    ? rendered
+    : elideSchema(schema, bytes);
```
[the remark this answers is not in the document]
Comment (mathpirate) [reply packages/state-inspector/decode.ts]: The observation is right and the docs were overclaiming, so I've fixed the claim. The patch I've not taken, because returning the raw schema breaks three things — probed rather than reasoned about:

- `JSON.stringify()` on a schema holding a `bigint` **throws** (`Do not know how to serialize a BigInt`). That call is the elision measurement, so every `annotate()` of such a link would crash.
- `stringifyInspectorJson()` on a raw schema **throws** on a `bigint` (`Inspector JSON contains an unannotated BigInt.`) and on a cycle (`Inspector JSON contains a circular structure.`). That is the `--full-depth` printer, so `return schema` crashes exactly the path the suggestion is trying to make faithful.
- A modern `FabricLink` under `default` raw-stringifies to `{}` and **disappears**. That is the vanishing act `decodedLinkOf()` exists to prevent, and it is the same class of defect as the one this PR fixes.

Annotating keeps the rendering JSON-safe and keeps one vocabulary across the output: a link inside a schema reads as `{ $link }` there the way it does everywhere else. A schema's own `$ref` is untouched by this — only a single-key `{ "/": … }` is rewritten — so the common case is unaffected, and there is now a test pinning that.

What I've changed instead, in ed11abf2a: `renderLinkSchema()` states that the rendering is the annotated form rather than the stored bytes, and why; the README and PR body say the same. The digest already hashes the **stored** schema rather than the rendering, so it remains the thing to compare two schemas by. Tests added for a schema's own `$ref`, for a sigil-shaped literal under `const`, and for a `bigint` schema rendering JSON-safely.

File: packages/state-inspector/detail.ts [hunk 2]
```diff
@@ -196,29 +197,25 @@ function parseCfc(cfc: unknown): CfcSummary | undefined {
...
-    const linkKey = Object.keys(slash).find((k) => k.startsWith("link@"));
-    const inner = linkKey ? slash[linkKey] : undefined;
-    if (isObjectNotArray(inner) && isObjectNotArray(inner.schema)) {
-      return {
-        schema: annotate(inner.schema),
-        keys: Object.keys(inner.schema),
-        via: "link",
-      };
-    }
+  const linkSchema = decodedLinkOf(naming)?.schema;
```
[the remark this answers is not in the document]
Comment (mathpirate) [reply packages/state-inspector/detail.ts:212]: Confirmed and fixed in ed11abf2a — `nameOf` now indexes by `decodedLinkOf(val)?.id`. You were right that the change was dead for modern links: with the sigil-only lookup the child is never named, so it never gets a label and never reaches `declaredSchemaFor()` at all. The PR body claimed otherwise and has been corrected.

`linksWithPaths()` had the same gap, so a modern link was also missing from a piece's outgoing-link list; that one reads either at-rest form now too. `legacyResultId()` keeps `parseSigilLink()`, since a legacy process cell's `resultRef` is sigil by construction.

Pinned by a test that seeds a piece whose value holds a real `FabricLink` (encoded through the codec with modern cell-rep on, which is the only way one survives at rest) and asserts the child is labeled `modernNamed`. It fails against the sigil-only lookup.

File: packages/state-inspector/detail.ts [hunk 3]
```diff
@@ -196,29 +197,25 @@ function parseCfc(cfc: unknown): CfcSummary | undefined {
...
+  if (isObjectNotArray(linkSchema)) {
+    return {
+      schema: annotate(linkSchema),
+      keys: Object.keys(linkSchema),
+      via: "link",
+    };
   }
```
[the remark this answers is not in the document]
Comment (mathpirate) [reply packages/state-inspector/detail.ts]: Agreed, and taken as suggested in ed11abf2a. I had preserved the object-only condition to keep the change tight, which was the wrong call: it is the same defect this PR is about. A link storing `true` says it carries no constraint, and falling through to the owner's declaration reports a constraint the link does not carry — substituting a different schema for the one value that means "there is none here". `false` is worse, since it admits nothing at all.

Any stored schema answers now, and `keys` is computed only for an object one. `via: "link"` still tells the reader which source won, so the owner's declaration is not hidden, just no longer allowed to impersonate the link's.

Pinned by a test with three named cells off one owner — link-with-object-schema, link-with-`true`, and link-with-none — asserting the third falls back to the owner's declaration and the second does not.

File: packages/state-inspector/decode.ts [hunk 4]
```diff
@@ -197,10 +204,74 @@ interface AnnotationLeave {
...
+ * like any other value, so a sigil-shaped literal inside one — under `const`,
+ * `default`, or `enum` — reads back as `{ $link }` or `{ $ref }` the way it
+ * would anywhere else in the output. That keeps the rendering JSON-safe, which
+ * the stored form is not: a `bigint` in a schema breaks `JSON.stringify()`
+ * outright, and a `FabricLink` in one flattens to `{}` and disappears. The
+ * digest in an `elideSchema()` summary hashes the stored schema rather than
+ * this rendering, so it stays the thing to compare two schemas by.
+ */
+function renderLinkSchema(schema: Json, maxDepth: number): Json {
+  const rendered = annotate(schema, Number.POSITIVE_INFINITY);
```
[the remark this answers is not in the document]
Comment (mathpirate) [reply packages/state-inspector/decode.ts]: Real, and fixed in 8e4456290 — thanks. `renderLinkSchema()` now measures the schema as stored and only builds the annotated copy when it is going to write one out. Benchmarked on the schema from the repro space (~2.2KB of `$defs`), 5000 schema-bearing links through `annotate()` at default depth: **193ms → 73ms**.

It also settles an inconsistency the summary had: `bytes` described the rendering while `digest` hashed the store, so the two halves of one marker were about different things. Both are about the stored schema now.

Two bounds, since measuring the stored form is not free of edge cases:

- `JSON.stringify()` refuses a `bigint` and refuses a cycle, so `jsonByteLength()` falls back to measuring the annotated form for those. Covered by a test with a `bigint` in the schema.
- A `FabricSpecialObject` stringifies to `{}` and so under-reports. The only consequence is that such a schema can be written out inline while rendering a little past the bound; it is stated in the doc comment.

The stored-vs-rendered distinction is pinned by a test using a sigil under `const`, which renders three bytes longer than it stores — it fails against the old measurement.

File: packages/state-inspector/detail.ts [hunk 5]
```diff
@@ -196,29 +197,28 @@ function parseCfc(cfc: unknown): CfcSummary | undefined {
...
-      return {
-        schema: annotate(inner.schema),
-        keys: Object.keys(inner.schema),
-        via: "link",
-      };
-    }
+  const linkSchema = decodedLinkOf(naming)?.schema;
+  if (linkSchema !== undefined) {
+    return {
+      schema: annotate(linkSchema),
```
[the remark this answers is not in the document]
Comment (mathpirate) [reply packages/state-inspector/detail.ts:215]: Not taking this one, though the size concern behind it is real.

The two positions have different jobs. Under `$link`, a schema is incidental to a value the operator asked for — they wanted the `topics` array, and kilobytes of `$defs` per element drowns it, which is what the elision is for. In `declaredSchemaFor()`, the schema *is* the answer: the field exists to say what a stream's payload or an owned cell's value type is declared to be, and eliding it there removes the thing the reader came for.

Consistency decides it too. The sibling branch of that same function resolves the owner's `schema.properties[<key>]` and renders it with `annotate()`, and `describeEntity()` renders an entity's own `doc.schema` the same way. Bounding only the link-sourced one would make two declarations rendered side by side, distinguishable in the display only by `via`, follow different rules.

No fidelity is at stake here either — `annotate()` at finite depth truncates with `"…"`, which is not mistakable for a schema. This is purely about output size.

The bundle-size point stands on its own, but it is pre-existing and symmetric: `doc.schema` is the far more common large schema in a bundle, since every piece carries a result schema, and it is unbounded today. If HTML bundle size wants bounding, it wants bounding across all three of those sites at once rather than at the one this PR happened to touch — worth its own change, and I have not filed one.

File: packages/state-inspector/decode.ts [hunk 6]
```diff
@@ -197,10 +204,98 @@ interface AnnotationLeave {
...
+ * schema: `true`, `false`, and every object shape are values a link can really
+ * store, so a summary wearing one of those would be indistinguishable from the
+ * thing it stands for. The wrapper is not proof against a consumer that treats
+ * whatever it finds here as JSON Schema and ignores keywords it does not know
+ * — nothing rendered in this position could be — but it cannot be mistaken for
+ * a stored schema by a reader who looks at it.
+ */
+function elideSchema(schema: Json, bytes: number): Json {
+  return {
+    $elidedSchema: {
```
Comment (mathpirate) [inline packages/state-inspector/decode.ts]: **[P2] Elision output can be indistinguishable from a real stored schema.**

This marker occupies the schema value slot, but a link can legitimately store an object having exactly this shape and content. I reproduced identical annotated output for a large elided schema and for the resulting marker stored as the real schema. Move the elision discriminator outside the schema value—such as a sibling `$link.schemaSummary` field—so output always distinguishes stored data from inspector metadata.

File: packages/state-inspector/decode.ts [hunk 7]
```diff
@@ -197,10 +204,98 @@ interface AnnotationLeave {
...
+ * The measurement is of the value as stored, which a `FabricSpecialObject`
+ * under-reports: it stringifies to `{}` whatever it holds. Such a schema can
+ * therefore be written out inline while rendering a little longer than the
+ * bound. Nothing downstream reads the count as an allocation size.
+ */
+function jsonByteLength(value: Json): number {
+  try {
+    return utf8.encode(JSON.stringify(value)).length;
+  } catch {
+    const rendered = JSON.stringify(annotate(value, Number.POSITIVE_INFINITY));
```
Comment (mathpirate) [inline packages/state-inspector/decode.ts]: **[P2] Schema measurement can overflow the stack.**

The fallback retries a failed recursive `JSON.stringify` with another recursive `JSON.stringify`. A valid 20,000-level schema parses successfully but causes default `annotate()` to throw `RangeError: Maximum call stack size exceeded` before it can emit an elision marker. Use stack-safe or bounded measurement and hashing, with a graceful fallback when an exact digest cannot be computed.

File: skills/state-inspector/SKILL.md [hunk 8]
```diff
@@ -115,6 +115,17 @@ Confusing an approximation for truth is the failure mode that matters here:
   legitimately differ. The scan labels `cross-space-linked` (real replica →
   drift bug) vs `no-cross-space-link` (likely independent instance). Don't cry
   wolf on the latter.
+- **A `schema` under `$link` is what the link stores, and a `$elidedSchema` is a
+  summary of it.** A link's schema is a JSON Schema, so `true` and `false` are
+  values it can really hold — `true` constrains nothing and lets the reader's
```
Comment (mathpirate) [inline skills/state-inspector/SKILL.md]: **[P2] Stored `true` currently replaces the reader schema.**

This is not current runtime behavior. `resolveLinkTracingDereferences()` carries the reader schema only when the stored schema is `undefined`; a literal stored `true` takes the other branch and replaces the reader schema. Remove this claim or defer it until the associated runtime change lands.

File: skills/state-inspector/SKILL.md [hunk 9]
```diff
@@ -115,6 +115,17 @@ Confusing an approximation for truth is the failure mode that matters here:
...
+  own schema travel through the hop, while a schema with properties in it
+  freezes a shape at that link. Never read one as the other. A schema too large
+  to print inline is replaced by `{ $elidedSchema: { keys, bytes,
+  digest } }`,
+  which is a marker rather than a schema: equal digests mean equal schemas, so
```
Comment (mathpirate) [inline skills/state-inspector/SKILL.md]: **[P3] Digest equality is probabilistic.**

The displayed digest is a truncated hash, so equal digests are strong evidence rather than proof of equal schemas. Say that different digests prove inequality, while equal digests make equality overwhelmingly likely and `--full-depth` is needed for certainty.

**Review by mathpirate — COMMENTED** [review]
Four findings from a full review are detailed inline: three P2 correctness/robustness issues and one P3 documentation-accuracy issue.

File: packages/state-inspector/decode.ts (hunk 6 above)
Comment (mathpirate) [reply packages/state-inspector/decode.ts]: You're right, and it's the same defect as the one this PR is about, one level in. I reproduced it before fixing:

```
elided marker:  {"$elidedSchema":{"keys":["type","description"],"bytes":434,"digest":"OvlEyImTp4hI"}}
that object stored as a real schema: {"$elidedSchema":{"keys":["type","description"],"bytes":434,"digest":"OvlEyImTp4hI"}}
INDISTINGUISHABLE: true
```

I had reached for the `$`-prefixed marker because it matches the file's existing vocabulary (`$link`, `$ref`, `$stream`, `$bigint`), and talked myself out of the sibling on those grounds. That was the wrong axis to optimize: an in-slot marker is only unforgeable for the stored values that happen not to collide with it, which is precisely the reasoning that made `schema: true` look acceptable.

Fixed in 34aef345f. The summary is now `$schemaSummary`, a sibling of `schema` on `$link`, and the two never both appear. What makes this hold for *every* stored value rather than for most is `payloadToLink()`: it bounds what a stored link contributes to `id`, `space`, `path`, `scope`, and `schema`, so nothing stored can reach a `$`-prefixed sibling.

Against the repro space:

```json
"$link": {
  "id": "of:fid1:1D7eZcK…",
  "scope": "space",
  "$schemaSummary": { "keys": ["$defs", "description", "properties", "required", "tags", "type"],
                      "bytes": 2206, "digest": "dP-TGNZVNhod" }
}
```

Pinned by a test that stores a summary verbatim as another link's schema and asserts it comes back under `schema` with no `$schemaSummary` beside it. It fails against the in-slot design.

File: packages/state-inspector/decode.ts (hunk 7 above)
Comment (mathpirate) [reply packages/state-inspector/decode.ts]: Real crash, fixed in 34aef345f — though the call responsible turns out to be a different one, which changed the fix.

I measured each call against a deeply nested schema:

```
 4000: JSON.stringify(raw)                ok
 4000: annotate(raw, Inf)                 ok
 4000: hashStringOf(raw)                  THROWS RangeError
 4000: stringifyInspectorJson(annotated)  ok
20000: (same)
```

`annotate()` is worklist-based and `stringifyInspectorJson()` was written non-recursive on purpose, so both survive; `JSON.stringify` is stack-safe in V8 and still succeeded at 200,000 levels. The recursion is in **`hashStringOf()`** — computing the digest. So `RangeError` was escaping from the summary, not from the measurement.

`schemaDigest()` now returns nothing rather than throwing, and the summary omits `digest`. The doc comment states the bound that matters: an absent digest means it was not computed, so two summaries that both lack one say nothing about whether their schemas agree. Test asserts a 20,000-level schema still yields `keys` and `bytes` with no `digest`.

One thing I did not keep: I first wrote `jsonByteLength()` to fall back a second time and return `undefined` when nothing could measure the value. That branch is unreachable — the only things `JSON.stringify` refuses are `bigint`s and cycles, and `annotate()` lowers both, so the fallback has nothing left to fail on however deep the value nests. It cost 2 uncovered lines on the ratchet and could not be tested, so I removed it rather than carry untestable defensive code.

File: skills/state-inspector/SKILL.md (hunk 8 above)
Comment (mathpirate) [reply skills/state-inspector/SKILL.md]: Correct, and this was the worst thing in the PR — a false claim about runtime behavior, in the skill that exists to be ground truth. Verified against main:

```ts
// packages/runner/src/link-resolution.ts:582
if (nextLink.schema === undefined && link.schema !== undefined) {
```

A stored `true` is not `undefined`, so it takes the `else` branch and replaces, exactly as you say. The behavior I described belongs to `schemaConstrainsNothing()` on `fable/array-read-schema-link` (#6059), which is unmerged — I had that branch's semantics in my head from reading it as related work and wrote it as though it were current.

Fixed in 34aef345f by removing the claim rather than deferring it. The skill and README now say what `true` means *as a schema* — it selects every value, so a link storing it declares no constraint, where a schema with properties in it pins a shape — and say nothing about how a hop resolves. That is true today and stays true when #6059 lands, so nothing here needs revisiting then.

The PR description carried the same sentence; it is corrected too. The premise is unaffected: `true` is still a legal stored value, still semantically distinct from a constraining schema, and rendering one as the other is still a fabrication. It just is not the resolution mechanism I attributed it to.

File: skills/state-inspector/SKILL.md (hunk 9 above)
Comment (mathpirate) [reply skills/state-inspector/SKILL.md]: Agreed — 12 base64url characters is 72 bits, which is evidence rather than proof, and "equal digests mean equal schemas" overclaimed it.

Fixed in 34aef345f, worded with the asymmetry you named: different digests prove two schemas differ; equal ones make agreement overwhelmingly likely without proving it, and `--full-depth` settles it. That wording is now in the skill, the README, and the doc comment on `schemaSummary()`, so the three do not drift apart.

The skill also picks up the neighbouring case, since it has the same shape: an absent `digest` means it could not be computed, and two summaries that both lack one say nothing about whether they agree.

File: skills/state-inspector/SKILL.md [hunk 10]
```diff
@@ -115,6 +115,21 @@ Confusing an approximation for truth is the failure mode that matters here:
...
+  value, declaring no constraint at all, where a schema with properties in it
+  pins a shape at that link. Never read one as the other. The summary is
+  `{ keys, bytes, digest }` and is a SIBLING of `schema`, never a value under
+  it, because a link can store a schema of any shape and a summary in the
+  `schema` slot could be a schema some link really holds. The two never both
+  appear: a `schema` key means that is the stored schema, a `$schemaSummary` key
+  means it was too large to print, and neither means the link stores no schema.
+  Different digests prove two schemas differ; equal ones make agreement
+  overwhelmingly likely without proving it, since the hash is truncated — reach
+  for `--full-depth` when you need certainty or the text itself. A `bytes` or
```
[the remark this answers is not in the document]
Comment (mathpirate) [reply skills/state-inspector/SKILL.md]: Correct on both counts — stale text I left behind. When I removed the unreachable second fallback from `jsonByteLength()`, I fixed the doc comment on `schemaSummary()` and missed the same sentence in the skill.

Fixed in the commit above. `bytes` is stated as always present; the two conditional fields are `digest`, absent when hashing a schema nested past the call stack, and `keys`, absent when the stored schema is not an object. Both the skill and the doc comment say it that way now.

`keys` is reachable rather than theoretical — stored data decides the shape, and a long string sitting where a schema should be is summarized without keys — so there is a test pinning it.
